# Data preparation

## Background

The dataset of consumer complaints on financial products was obtained from [Kaggle](https://www.kaggle.com/ashwinik/consumer-complaints-financial-products?select=Consumer_Complaints.csv). 

This data is a collection of complaints about consumer financial products and services that we sent to companies for response. Complaints are published after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first.

## Import packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime

## Read in dataset

Column datatypes (taken from the descriptions on the Kaggle website.)

The 'ZIP code' column contains NAN entries so read in as a string for now, and convert to int type later. (Actually not a relevant column so will end up deleting it anyway.)

In [2]:
data_types = {'Date received': str, 'Product': str, 'Sub-product': str, 'Issue': str, 'Sub-issue': str,
       'Consumer complaint narrative': str, 'Company public response': str, 'Company': str,
       'State': str, 'ZIP code': str, 'Tags': str, 'Consumer consent provided?': str,
       'Submitted via': str, 'Date sent to company': str, 'Company response to consumer': str,
       'Timely response?': str, 'Consumer disputed?': str, 'Complaint ID': int}

Date columns to be parsed.

In [3]:
parse_dates = ['Date received', 'Date sent to company']

Read in data.

In [4]:
complaints_df = pd.read_csv('~/documents/data/consumer_complaints/consumer_complaints_raw.csv', \
                            dtype = data_types, parse_dates = parse_dates)

## Shape

In [5]:
print(complaints_df.shape)

(670598, 18)


## Check for duplicate rows

In [6]:
print(complaints_df.duplicated().sum())

0


## Check for null or NA values

In [7]:
print(complaints_df.isna().sum(), '\n')

Date received                        0
Product                              0
Sub-product                     198202
Issue                                0
Sub-issue                       400730
Consumer complaint narrative    555894
Company public response         525401
Company                              0
State                             5305
ZIP code                          5324
Tags                            575868
Consumer consent provided?      462447
Submitted via                        0
Date sent to company                 0
Company response to consumer         0
Timely response?                     0
Consumer disputed?               41419
Complaint ID                         0
dtype: int64 



Keep only rows with 'Consumer complaint narrative' provided.

In [8]:
complaints_df = complaints_df[complaints_df['Consumer complaint narrative'].notnull()]

In [9]:
print(complaints_df.shape)

(114704, 18)


In [10]:
print(complaints_df.isna().sum(), '\n')

Date received                       0
Product                             0
Sub-product                     35784
Issue                               0
Sub-issue                       57448
Consumer complaint narrative        0
Company public response         57417
Company                             0
State                             348
ZIP code                          353
Tags                            95333
Consumer consent provided?          0
Submitted via                       0
Date sent to company                0
Company response to consumer        0
Timely response?                    0
Consumer disputed?               8421
Complaint ID                        0
dtype: int64 



## Drop irrelevant columns

In [11]:
complaints_df = complaints_df.drop(['Date received', 'Company public response', 'State', 'ZIP code', 'Tags',\
                                   'Consumer consent provided?', 'Submitted via', 'Date sent to company',\
                                   'Company public response', 'Timely response?', 'Consumer disputed?', \
                                    'Complaint ID', 'Company response to consumer'], axis = 1)

In [12]:
print(complaints_df.shape)

(114704, 6)


In [13]:
print(complaints_df.isna().sum(), '\n')

Product                             0
Sub-product                     35784
Issue                               0
Sub-issue                       57448
Consumer complaint narrative        0
Company                             0
dtype: int64 



## Check for duplicate rows

In [14]:
print(complaints_df.duplicated().sum())

746


746 is a very small proportion of the 114,704 rows, so OK to drop duplicates.

In [15]:
complaints_df = complaints_df.drop_duplicates()

## Save clean data to file

In [17]:
complaints_df.to_csv('~/documents/data/consumer_complaints/consumer_complaints_clean.csv')